framework/src/main/java/org/checkerframework/common/value/ValueVisitor.java

```diff
@@ -324,6 +333,154 @@ public Void visitTypeCast(TypeCastTree tree, Void p) {
     return super.visitTypeCast(tree, p);
   }
 
+  // At this point, types are like: (@IntVal(-1) byte, @IntVal(255) int) and knowledge of signedness
+  // is gone.  So, use castType's underlying type to infer correctness of the cast.  This method
+  // returns true for (@IntVal(-1), @IntVal(255)) if the underlying type is `byte`, but not for any
+  // other underlying type.
+  @Override
+  protected boolean isTypeCastSafe(AnnotatedTypeMirror castType, AnnotatedTypeMirror exprType) {
+    TypeKind castTypeKind = TypeKindUtils.primitiveOrBoxedToTypeKind(castType.getUnderlyingType());
+    TypeKind exprTypeKind = TypeKindUtils.primitiveOrBoxedToTypeKind(exprType.getUnderlyingType());
+    if (castTypeKind != null
+        && exprTypeKind != null
+        && TypeKindUtils.isIntegral(castTypeKind)
+        && TypeKindUtils.isIntegral(exprTypeKind)) {
+      AnnotationMirrorSet castAnnos = castType.getAnnotations();
+      AnnotationMirrorSet exprAnnos = exprType.getAnnotations();
+      if (castAnnos.equals(exprAnnos)) {
+        return true;
+      }
+      assert castAnnos.size() == 1;
+      assert exprAnnos.size() == 1;
+      AnnotationMirror castAnno = castAnnos.first();
+      AnnotationMirror exprAnno = exprAnnos.first();
+      boolean castAnnoIsIntVal = atypeFactory.areSameByClass(castAnno, IntVal.class);
+      boolean exprAnnoIsIntVal = atypeFactory.areSameByClass(exprAnno, IntVal.class);
+      if (castAnnoIsIntVal && exprAnnoIsIntVal) {
+        List<Long> castValues = atypeFactory.getIntValues(castAnno);
+        List<Long> exprValues = atypeFactory.getIntValues(exprAnno);
+        if (castValues.size() == 1 && exprValues.size() == 1) {
+          // Special-case singleton sets for speed.
+          switch (castTypeKind) {
+            case BYTE:
+              return castValues.get(0).byteValue() == exprValues.get(0).byteValue();
+            case INT:
+              return castValues.get(0).intValue() == exprValues.get(0).intValue();
+            case SHORT:
+              return castValues.get(0).shortValue() == exprValues.get(0).shortValue();
+            default:
+              return castValues.get(0).longValue() == exprValues.get(0).longValue();
+          }
+        } else {
+          switch (castTypeKind) {
+            case BYTE:
+              {
+                TreeSet<Byte> castValuesTree =
+                    new TreeSet<Byte>(CollectionsPlume.mapList(Number::byteValue, castValues));
+                TreeSet<Byte> exprValuesTree =
+                    new TreeSet<Byte>(CollectionsPlume.mapList(Number::byteValue, exprValues));
+                return sortedSetContainsAll(castValuesTree, exprValuesTree);
```

Comment:
These sets are no more than 10 elements long, so it's not really necessary to use this special method.

Reply:
For sets of size 4, the method saves over 10% of run time.  That said, I do agree this isn't a bottleneck for the checker.